File: impl/sdk/java/src/examples/java/io/resonatehq/examples/humanintheloop/HumanInTheLoop.java
package io.resonatehq.examples.humanintheloop;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.resonatehq.resonate.Context;
import io.resonatehq.resonate.Context.ResonateFuture;
import io.resonatehq.resonate.Handle.ResonateHandle;
import io.resonatehq.resonate.Resonate;
import io.resonatehq.resonate.Types.Value;
import java.util.concurrent.CompletableFuture;

/**
 * human shows a human-in-the-loop durable workflow.
 *
 * <p>An order-fulfillment orchestrator does some prep work, then <i>suspends</i> on a durable promise
 * that some external party -- a reviewer, a webhook, a UI -- is expected to resolve. While the
 * workflow is suspended, the worker holds no state: the orchestrator can be restarted, the server can
 * be restarted, even days can pass. Whenever the external resolve eventually arrives, replay picks up
 * exactly where it left off and proceeds with the human's decision.
 *
 * <p>The mechanism is {@link Context#promise()}: a "dependency-injected" durable promise with a
 * global, externally addressable id. The orchestrator awaits it; anyone with the id can settle it
 * through the regular promise API ({@code r.promises.resolve(id, ...)}), the CLI, or HTTP.
 *
 * <pre>{@code ./gradlew runExample -PmainClass=io.resonatehq.examples.humanintheloop.HumanInTheLoop                        # approve
 * ./gradlew runExample -PmainClass=io.resonatehq.examples.humanintheloop.HumanInTheLoop -PexampleArgs="--decision reject"}</pre>
 *
 * <p>Note on replay: the orchestrator re-executes from the top on every suspend, so any side effect
 * (printing the promise id, calling a notification service) belongs in a leaf -- {@code
 * notifyReviewer} here -- which settles once and never re-runs.
 */
public final class HumanInTheLoop {
    private HumanInTheLoop() {}

    private static final ObjectMapper MAPPER = new ObjectMapper();

    // -- Reviewer inbox --------------------------------------------------------

    /**
     * Signal channel from the worker to a waiting reviewer.
     *
     * <p>{@code notifyReviewer} is the leaf that <i>publishes</i> the durable promise id (in a real
     * system: to Slack, email, a dashboard). Here it also completes {@code approvalId} so anything
     * awaiting an approval -- like the simulated reviewer in {@code main} -- learns the id the instant
     * the promise exists, rather than polling until it stops 404-ing.
     *
     * <p>Injected through {@link Resonate#withDependency(Object)} and fetched in the leaf via {@code
     * ctx.getDependency(ReviewerInbox.class)}.
     */
    public static final class ReviewerInbox {
        final CompletableFuture<String> approvalId = new CompletableFuture<>();

        public void publish(String approvalId) {
            // notifyReviewer is a leaf and settles once, but be defensive: complete only once.
            this.approvalId.complete(approvalId);
        }
    }

    // -- Domain types ----------------------------------------------------------

    public record Decision(boolean approve, String note) {}

    // -- Leaf functions (each prints once, settles once) -----------------------

    public static String notifyReviewer(Context ctx, String orderId, int amount, String approvalId) {
        // In a real system this would call Slack / email / a dashboard. The point is that the side
        // effect (and the promise id the reviewer must resolve) lives in a leaf, not in the
        // orchestrator -- so it happens exactly once.
        System.out.printf(
                "  [notify_reviewer] order %s ($%d) needs approval; resolve promise id: '%s'%n",
                orderId, amount, approvalId);
        // Signal any in-process listener (the simulated reviewer in main) that the durable promise now
        // exists and tell them its id. A real reviewer would learn the same id from the side effect.
        ctx.getDependency(ReviewerInbox.class).publish(approvalId);
        return approvalId;
    }

    public static String shipOrder(Context ctx, String orderId, String note) {
        System.out.printf("  [ship_order] shipping %s (note: '%s')%n", orderId, note);
        return "shipped-" + orderId;
    }

    public static String cancelOrder(Context ctx, String orderId, String note) {
        System.out.printf("  [cancel_order] canceling %s (reason: '%s')%n", orderId, note);
        return "canceled-" + orderId;
    }

    // -- Orchestrator ----------------------------------------------------------

    public static String fulfillOrder(Context ctx, String orderId, int amount) {
        // Open the human-decision promise first so its id is deterministic ({workflow_id}.1).
        // ctx.promise returns a future whose id() is awaitable; we publish that id through a leaf so a
        // real reviewer would know where to resolve.
        ResonateFuture<Object> approval = ctx.promise(); // inherit workflow timeout
        String approvalId = approval.id();

        ctx.run(HumanInTheLoop::notifyReviewer, orderId, amount, approvalId).await();

        // Suspend until the external party resolves the promise. The worker holds no state while
        // suspended; this can be seconds, hours, or days.
        Decision decision = MAPPER.convertValue(approval.await(), Decision.class);

        if (decision.approve()) {
            return ctx.run(HumanInTheLoop::shipOrder, orderId, decision.note()).await();
        }
        return ctx.run(HumanInTheLoop::cancelOrder, orderId, decision.note()).await();
    }

    // -- main ------------------------------------------------------------------

    /**
     * Stand in for an external system that eventually resolves the promise. Waits for {@code
     * notifyReviewer} to publish the durable promise id on the inbox -- exactly mirroring how a real
     * reviewer learns where to resolve -- then settles it. No polling, no hardcoded id.
     */
    private static void simulateReviewer(Resonate r, ReviewerInbox inbox, Decision decision) {
        String approvalId = inbox.approvalId.join();
        r.promises.resolve(approvalId, new Value(null, decision)).join();
        System.out.printf(
                "[reviewer] resolved %s -> approve=%b note='%s'%n", approvalId, decision.approve(), decision.note());
    }

    public static void main(String[] args) {
        String decisionArg = argValue(args, "--decision", "approve");
        String order = argValue(args, "--order", "order-42");
        int amount = Integer.parseInt(argValue(args, "--amount", "199"));

        String url = System.getenv().getOrDefault("RESONATE_URL", "http://localhost:8001");
        Resonate r = Resonate.builder().url(url).build();
        // Hand the inbox to every context -- notifyReviewer will publish on it the moment the durable
        // approval promise is created.
        ReviewerInbox inbox = new ReviewerInbox();
        r.withDependency(inbox);
        r.register(HumanInTheLoop::fulfillOrder);

        try {
            String wid = "fulfill-" + System.nanoTime();
            System.out.println("[fulfill_order] starting workflow id=" + wid + " decision='" + decisionArg + "'");
            ResonateHandle<String> handle = r.run(wid, HumanInTheLoop::fulfillOrder, order, amount);

            boolean approve = decisionArg.equals("approve");
            Decision decision = new Decision(approve, approve ? "looks good" : "policy violation");
            Thread reviewer = new Thread(() -> simulateReviewer(r, inbox, decision));
            reviewer.start();

            String out = handle.result();
            try {
                reviewer.join();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            if (approve) {
                assert "shipped-order-42".equals(out);
            } else {
                assert "canceled-order-42".equals(out);
            }
            System.out.println("[fulfill_order] OK: " + out);
        } finally {
            r.stop().join();
        }
    }

    private static String argValue(String[] args, String flag, String fallback) {
        for (int i = 0; i + 1 < args.length; i++) {
            if (args[i].equals(flag)) {
                return args[i + 1];
            }
        }
        return fallback;
    }
}

File: impl/sdk/ts/examples/generator/human-in-the-loop.ts
// human-in-the-loop shows a durable workflow that suspends on an external
// decision.
//
// An order-fulfillment orchestrator does some prep work, then *suspends* on a
// durable promise that some external party -- a reviewer, a webhook, a UI -- is
// expected to resolve. While suspended, the worker holds no state: the
// orchestrator, the server, even days, can pass. Whenever the external resolve
// arrives, replay picks up exactly where it left off.
//
// The mechanism is ctx.promise(): a durable promise with a global,
// externally-addressable id. The orchestrator awaits it; anyone with the id can
// settle it through the regular promise API (resonate.promises.resolve), the
// CLI, or HTTP.
//
// Start a Resonate server on localhost:8001 first (`resonate dev`), then:
//
//   npx tsx examples/generator/human-in-the-loop.ts                    # approve
//   npx tsx examples/generator/human-in-the-loop.ts --decision reject
//
// Note on replay: the orchestrator re-executes from the top on every suspend,
// so any side effect (publishing the promise id, calling a notifier) belongs in
// a leaf -- `notifyReviewer` here -- which settles once and never re-runs.

import { Codec, type Context, Resonate } from "../../src/index.js";

interface Decision {
  approve: boolean;
  note: string;
}

// Signal channel from the worker to a waiting reviewer. `notifyReviewer`
// publishes the durable promise id here; the simulated reviewer in main() awaits
// it, exactly mirroring how a real reviewer would learn where to resolve.
class ReviewerInbox {
  private resolve!: (id: string) => void;
  readonly approvalId = new Promise<string>((r) => {
    this.resolve = r;
  });
  private done = false;
  publish(id: string): void {
    if (!this.done) {
      this.done = true;
      this.resolve(id);
    }
  }
}

// -- Leaf functions (each prints once, settles once) ----------------------

async function notifyReviewer(ctx: Context, orderId: string, amount: number, approvalId: string): Promise<string> {
  // In a real system this would call Slack / email / a dashboard. The side
  // effect (and the promise id the reviewer must resolve) lives in a leaf.
  console.log(
    `  [notifyReviewer] order ${orderId} ($${amount}) needs approval; resolve promise id: ${JSON.stringify(approvalId)}`,
  );
  ctx.getDependency<ReviewerInbox>("inbox")!.publish(approvalId);
  return approvalId;
}

async function shipOrder(_ctx: Context, orderId: string, note: string): Promise<string> {
  console.log(`  [shipOrder] shipping ${orderId} (note: ${JSON.stringify(note)})`);
  return `shipped-${orderId}`;
}

async function cancelOrder(_ctx: Context, orderId: string, note: string): Promise<string> {
  console.log(`  [cancelOrder] canceling ${orderId} (reason: ${JSON.stringify(note)})`);
  return `canceled-${orderId}`;
}

// -- Orchestrator ---------------------------------------------------------

function* fulfillOrder(ctx: Context, orderId: string, amount: number): Generator<any, string, any> {
  // Open the human-decision promise first so its id is deterministic
  // (`{workflowId}.0`). The Future's `.id` is the address to resolve.
  const approval = yield* ctx.promise<Decision>();
  const approvalId = approval.id;

  yield* ctx.run(notifyReviewer, orderId, amount, approvalId);

  // Suspend until the external party resolves the promise. The worker holds no
  // state while suspended; this can be seconds, hours, or days.
  const decision: Decision = yield* approval;

  if (decision.approve) return yield* ctx.run(shipOrder, orderId, decision.note);
  return yield* ctx.run(cancelOrder, orderId, decision.note);
}

// -- main -----------------------------------------------------------------

function flag(name: string, fallback: string): string {
  const i = process.argv.indexOf(`--${name}`);
  return i >= 0 && process.argv[i + 1] ? process.argv[i + 1] : fallback;
}

const approve = flag("decision", "approve") === "approve";

const resonate = new Resonate({ url: process.env.RESONATE_URL ?? "http://localhost:8001" });
const inbox = new ReviewerInbox();
resonate.setDependency("inbox", inbox);
resonate.register("fulfillOrder", fulfillOrder);

const codec = new Codec();

// Stand in for an external system that eventually resolves the promise: wait
// for notifyReviewer to publish the id, then settle it. No polling, no
// hardcoded id.
async function simulateReviewer(decision: Decision): Promise<void> {
  const approvalId = await inbox.approvalId;
  await resonate.promises.resolve(approvalId, { data: codec.encode(decision).data });
  console.log(`[reviewer] resolved ${approvalId} -> approve=${decision.approve} note=${JSON.stringify(decision.note)}`);
}

try {
  const wid = `fulfill-${Date.now()}`;
  console.log(`[fulfillOrder] starting workflow id=${wid} decision=${JSON.stringify(approve ? "approve" : "reject")}`);
  const decision: Decision = { approve, note: approve ? "looks good" : "policy violation" };
  const reviewer = simulateReviewer(decision);

  const out = await resonate.run(wid, fulfillOrder, "order-42", 199);
  await reviewer;
  const expected = approve ? "shipped-order-42" : "canceled-order-42";
  if (out !== expected) throw new Error(`expected ${expected}, got ${out}`);
  console.log(`[fulfillOrder] OK: ${out}`);
} finally {
  await resonate.stop();
}
